### ISSUE TITLE
Bug: Sequential.fit fails for Embedding → RNN → Dense with softmaxCrossEntropy

### ISSUE THREAD

**Issue by Akhyar11** [Akhyar11/ML-V1#41]
## Summary

`Sequential.fit()` fails when training a model composed of:

```ts
Embedding → RNN → Dense
```

The input data is already padded and has the expected shape `[seqLen, 1]`, but training fails during `Dense.backward()` with a dot product dimension mismatch.

## Environment

- Package: `@akhyar11/ml-v1`
- Runtime: Node.js via `ts-node`
- OS/path example: Linux
- Model type: text classification using BPE tokenizer + Embedding + RNN + Dense
- Loss: `softmaxCrossEntropy`
- Optimizer: `adam`

## Model configuration

```ts
import { BPETokenizer, RNN, Sequential, Embedding, Dense, mj, Matrix } from "@akhyar11/ml-v1";
import path from "path";
import { getDataset } from "./get_dataset";

const SEQ_LEN = 64;
const EMBEDDING_DIM = 64;
const HIDDEN_UNITS = 128;
const OUTPUT_CLASSES = 2;
const alpha = 1e-3;

const tokenizer = BPETokenizer.load(path.join(__dirname, "../tokenizer.json"));

const VOCAB_SIZE = 2685; // tokenizer log: 2685 tokens

const model = new Sequential({
  layers: [
    new Embedding({
      vocabSize: VOCAB_SIZE,
      embeddingDim: EMBEDDING_DIM,
      alpha,
    }),
    new RNN({
      units: EMBEDDING_DIM,
      hiddenUnits: HIDDEN_UNITS,
      activation: "tanh",
      returnSequences: false,
      stateful: false,
      alpha,
    }),
    new Dense({
      units: HIDDEN_UNITS,
      outputUnits: OUTPUT_CLASSES,
      activation: "linear",
      alpha,
    }),
  ],
});

model.compile({
  alpha,
  error: "softmaxCrossEntropy",
  optimizer: "adam",
  clipGradient: 5.0,
});
```

## Dataset preprocessing

The input sequence is padded to fixed length `64` and converted to matrix shape `[64, 1]`:

```ts
const x_token: Matrix[] = dataset.map(item => {
  const ids = tokenizer.encode(String(item.text ?? "").trim());
  const padded = tokenizer.padSequence(ids, SEQ_LEN);
  return mj.matrix(padded.map(id => [id]));
});

const y: Matrix[] = dataset.map(item => {
  const label = item.label === "positive" ? 1 : 0;
  return mj.matrix([[label]]);
});
```

Debug output confirms the first sample shape is correct:

```txt
dataset length: 700
x_token length: 700
first x shape: [ 64, 1 ]
first y shape: [ 1, 1 ]
first tokens: [
     4,  5,  413,   10,  415,
    16,  4,   17, 1354, 1355,
  1745, 10,    5,  106,   10,
   100, 16, 1746,  207, 1123
]
```

## Training call

```ts
model.fit(x_token, y, 10, {
  batchSize: 1,
  earlyStoppingPatience: 10,
  shuffle: true,
  verbose: true,
  validationSplit: 0.2,
});
```

`batchSize: 1` is used because recurrent layers currently require sample-by-sample training in generic `Sequential.fit()`.

## Actual behavior

Training fails at the beginning of epoch 1:

```txt
Epoch 1/10 [>                              ] 0.0% | Loss: .... | Val Loss: .... | 0.0 samples/s | ETA: --:--
Error: Dimensi matrix tidak cocok untuk dot product: [1x0] * [1x128]
    at Object.dotProduct (.../node_modules/@akhyar11/ml-v1/src/math/dotProduct.ts:31:11)
    at Dense.backward (.../node_modules/@akhyar11/ml-v1/src/layers/dense.ts:356:23)
    at Sequential.backward (.../node_modules/@akhyar11/ml-v1/src/models/sequential.ts:76:28)
    at Sequential.fit (.../node_modules/@akhyar11/ml-v1/src/models/sequential.ts:267:14)
    at Object.<anonymous> (.../src/main.ts:50:7)
```

Model summary before training:

```txt
========== Model Info ==========
Layer name   : embedding layer
Layer input  : [0,0]
Layer output : [0,0]
Layer param  : 171840

Layer name   : rnn layer
Layer input  : [64,0]
Layer output : [128,1]
Layer param  : 24704

Layer name   : dense layer
Layer input  : [128,1]
Layer output : [2,1]
Layer param  : 258

Total params = 196802
========== End Info ==========
```

## Expected behavior

`Sequential.fit()` should train successfully for this valid recurrent pipeline:

```txt
Input token matrix: [64, 1]
Embedding output:   [64, 64]
RNN output:         [128, 1]
Dense output:       [2, 1]
Target:             [1, 1]
```

Expected result: epoch training should proceed without dimension mismatch.

## Notes / suspected area

The input matrix shape is valid and non-empty. The error suggests that somewhere in the `Sequential.fit()` / `Sequential.backward()` path, the output/cache passed between `Embedding`, `RNN`, and `Dense` becomes empty along the sequence dimension, causing `Dense.backward()` to receive or use a cached input with zero columns.

Possible areas to inspect:

- `Sequential.forward()` / `Sequential.backward()` integration for `Embedding → RNN → Dense`
- Whether `Embedding.forward()` output orientation is compatible with `RNN.forward()` inside `Sequential`
- Whether `Dense.backward()` cache input is valid after RNN forward
- Whether recurrent layer handling in generic `Sequential.fit()` needs a special path for sequence layers

## Workaround

A possible temporary workaround is to avoid `model.fit()` and train manually:

```ts
const xEmb = embedding.forward(x);
const h = rnn.forward(xEmb);
const logits = dense.forward(h);

const denseErr = dense.backward(y, mj.matrix([[]]));
const rnnErr = rnn.backward(mj.matrix([[]]), denseErr);
embedding.backward(mj.matrix([[]]), rnnErr);
```

But ideally, `Sequential.fit()` should support this pipeline when `batchSize: 1` is used.

**Comment by Akhyar11** [comment]
Update: manual training loop works as a workaround.

I tested the same `Embedding → RNN → Dense` architecture without `Sequential.fit()`, using explicit forward/backward calls:

```ts
const xEmb = embedding.forward(sample.x);
const h = rnn.forward(xEmb);
const logits = dense.forward(h);

const denseErr = dense.backward(sample.y, emptyErr);
const rnnErr = rnn.backward(emptyErr, denseErr);
embedding.backward(emptyErr, rnnErr);
```

The input and intermediate shapes are valid:

```txt
x: [ 64, 1 ]
embedding output: [ 64, 64 ]
rnn output: [ 128, 1 ]
dense output: [ 2, 1 ]
```

Training ran successfully for 10 epochs with shuffled training samples each epoch and the model was saved successfully:

```txt
Epoch 1/10 selesai | Train Loss: 0.8024 | Train Acc: 48.39% | Val Acc: 52.14% | Time: 5.39s
...
Epoch 10/10 selesai | Train Loss: 0.7125 | Train Acc: 58.04% | Val Acc: 50.71% | Time: 5.68s
Model saved to: .../model.json
```

This suggests the layer implementations themselves can run this pipeline correctly, and the bug is likely isolated to the `Sequential.fit()` / `Sequential.backward()` integration path for `Embedding → RNN → Dense`.

Current workaround:

- Avoid `model.fit()` for this architecture.
- Use manual loop:
  - `Embedding.forward()`
  - `RNN.forward()`
  - `Dense.forward()`
  - `Dense.backward()`
  - `RNN.backward()`
  - `Embedding.backward()`
- Shuffle the training sample array manually at the beginning of each epoch.

Given this workaround is confirmed, I will close this issue from the user side unless we want to keep it open specifically to track a future fix for `Sequential.fit()` support.
